### DISCUSSION TITLE
How to profile a nicegui app?

### DISCUSSION THREAD

**Discussion by Karim-Ghibli** [zauberzeug/nicegui#2533] [Q&A]
### Question

Is there a way to profile a nicegui app?

I have a small personal use app and it has some serious performance issues and wanted to see which functions (whether some ui component creation/actions or maybe my own code of the data model) were slowing it down the most. 

I am mostly a Python guy, so what I have tried so far is just using cProfile, from command line and from the script itself like so:
```python
from nicegui import ui
import cProfile

ui.label("zzz")

cProfile.run("ui.run()")
```

And of course that didn't work, with "NameError: name 'ui' is not defined".

I also tried 

```python
from nicegui import ui
import cProfile

ui.label('zzz')

z = """
if __name__ in {"__main__", "__mp_main__"}:
    ui.run(port=8000)
"""
cProfile.run(z)
```

and 

```python
from nicegui import ui
import cProfile

ui.label("zzz")

if __name__ in {"__main__", "__mp_main__"}:
    cProfile.run("ui.run(port=8000)")
```

and those didn't work either.

I know that basically this is a framework so it's not as if I am directly executing the code, but I was hoping that there was a relatively easy way to profile nicegui apps (even if not with cProfile)?

**Comment by falkoschindler** [comment]
Hi @Karim-Ghibli,

I'm not sure if profiling `ui.run` is possible. But I can share two approaches we are using to profile parts of our applications:

1. With yappi you can start and stop the profiler and get an overview of all functions that have been called in between:
    ```py
    import tabulate
    import yappi
    from nicegui import ui
    
    def start() -> None:
        yappi.clear_stats()
        yappi.start()
    
    def stop() -> None:
        yappi.stop()
        table = [
            [str(v) for v in [stat.full_name, stat.ttot, stat.tsub, stat.tavg, stat.ncall]]
            for stat in yappi.get_func_stats()
            if 'python' not in stat.module
        ]
        print(tabulate.tabulate(table[:15], headers=['function', 'total', 'excl. sub', 'avg', 'ncall'], floatfmt='.4f'))
        yappi.get_thread_stats().print_all()
    
    ui.button('start', on_click=start)
    ui.button('stop', on_click=stop)

    ui.run()
    ```

2. For a more specific analysis you can decorate individual functions with a line profiler:
    ```py
    from line_profiler import LineProfiler
    from nicegui import ui
    
    profile = LineProfiler()
    
    @profile
    def compute():
        s = 0
        for i in range(1_000_000):
            s += i
        return s
    
    def run():
        compute()
        profile.print_stats()
    
    ui.button('run', on_click=run)

    ui.run()
    ```

**Comment by ed2050** [comment]
@falkoschindler Is there a way to disable multiprocessing in ui.run just for profiling?  I want to profile my app startup times.  I don't want yappi or LineProfiler, I don't want new third party modules.  I just want basic built-in python profiling that I already know how to use.

# Attempts So Far

I tried something similar to OP.  Launching my app with `python3 -m cProfile -o out.prof myapp.py` results in error message below (same with profile instead of cProfile).  The message is erroneous because
1. myapp.py already uses `if __name__ in {"__main__", "__mp_main__"}:`
2. myappy.py launches ui.run just fine this way without -m cProfile

The error could be multiprocessing related.  I disable multiprocessing in my own apps when I need to profile.

# Error message
```
ERROR:    Traceback (most recent call last):
  File "/Users/foobar/Library/Python/3.8/lib/python/site-packages/starlette/routing.py", line 677, in lifespan
    async with self.lifespan_context(app) as maybe_state:
  File "/Users/foobar/Library/Python/3.8/lib/python/site-packages/starlette/routing.py", line 566, in __aenter__
    await self._router.startup()
  File "/Users/foobar/Library/Python/3.8/lib/python/site-packages/starlette/routing.py", line 656, in startup
    handler()
  File "/Users/foobar/Library/Python/3.8/lib/python/site-packages/nicegui/nicegui.py", line 75, in handle_startup
    raise RuntimeError('\n\n'
RuntimeError: 

You must call ui.run() to start the server.
If ui.run() is behind a main guard
   if __name__ == "__main__":
remove the guard or replace it with
   if __name__ in {"__main__", "__mp_main__"}:
to allow for multiprocessing.

ERROR:    Application startup failed. Exiting.
```

**Reply by falkoschindler** [reply]
The problem might be related to #1993, which we haven't found a solution for. Maybe running with `reload=False` helps?

**Reply by ed2050** [reply]
That's so funny.  I was just experimenting and discovered that 
- most of the profile time in my attempt above was actually spent in `watch` module
- cProfile actually works with `reload=False`
- startup is noticeably faster with reload off

So yes, I can confirm.  `cProfile` works when `reload=False`!  Maybe because cprofile can't handle the rust code in watch module?
